### ISSUE TITLE
Compiler Plugin COMPILATION_ERROR: Unresolved reference

### ISSUE THREAD

**Issue by Tree4Free** [tschuchortdev/kotlin-compile-testing#40]
When having a compilerPlugin registered that generates a method into classes annotated with a specific annotation its not possible to call the method in the code that should be compiled because it complains that the method call is an Unresolved reference.

Given the following code to compile:
```kotlin
@GenerateMethod
class ExampleClass {
    fun callGeneratedMethod() {
        this.generatedMethod() // <-- causes "Unresolved reference: generatedMethod"
    }
}
```

The method is generated via an ExpressionCodegenExtension and is also visible and callable using reflection. (To see this I removed the call to `this.generatedMethod()` which then compiles).

It seems like the error is thrown even before my compiler plugin is executed so my plugin has no chance of generating the method.

**Comment by tschuchortdev** [comment]
Can you give a sample project to reproduce the problem? 

> It seems like the error is thrown even before my compiler plugin is executed so my plugin has no chance of generating the method.

Sounds like the issue could be with execution order of the plugins. 

Does the plugin work correctly in a regular Gradle project?

**Comment by Tree4Free** [comment]
I am sorry the issue was on my end. While creating a sample project I discovered that I was using an older way (at least it compiled like this back in the day 😄 ) of defining the methods using only an `ExpressionCodegenExtension` (I think this was kotlin 1.3.50).

Seems like now with kotlin 1.3.71 you have to also register a `SyntheticResolveExtension` that first provides a synthetic `SimpleFunctionDescriptor` and then in the `ExpressionCodegenExtension` I am able to generate an implementation for that method.
Doing it that way it now works from both maven as well as from your testing library.

Maybe you could help me with one additional little detail: 

When running the compiler plugin with maven, i have to use the `kotlin-maven-plugin` which pulls in the `kotlin-compiler`, which results in me using the ComponentRegistrar from that dependency. 

The `kotlin-compiler-embeddable` pulled in by your library also has a ComponentRegistrar however the MockProject is in another package 

(`com.intellij.mock.MockProject` in compiler vs `org.jetbrains.kotlin.com.intellij.mock.MockProject` in compiler-embeddable)  

So it seems like I am unable to use maven and your library at the same time, since either way the other will complain that the plugin is not compatible, do you know a way around that?

Sample project is here: [compiler-plugin-test-sample](https://github.com/Tree4Free/compiler-plugin-test-sample)

The kotlin compiler dependency is specified in this file: `/compiler-plugin/pom.xml`

Thanks for your help, you can close this issue. 😅

**Comment by tschuchortdev** [comment]
Why are dependencies that are only needed for the maven plugin available during runtime? Those classpaths should be separate imo.

One thing you could try is to override/exclude the `kotlin-compiler-embeddable` dependency of kotlin-compile-testing with the `kotlin-compiler` dependency. As long as kotlin-compile-testing doesn't use any API directly that is different between the two, there's a good chance that it could work. 

Please report back if you find a solution so I can add it to the readme. 

This dependency nightmare has been bugging me a lot but I have no idea how it could be resolved. It would be nice if kotlin-compile-testing could be used with different compiler versions at runtime as well.

**Comment by Tree4Free** [comment]
> Why are dependencies that are only needed for the maven plugin available during runtime? Those classpaths should be separate imo.

This lead me onto the right track!

I went back and analyzed how exactly this is done with the official compiler plugins and it turns out there are two versions for the compiler plugin, one for maven and one for gradle and the key is that they use a different `ComponentRegistrar`.

- The maven plugin uses the `ComponentRegistrar` from the `kotlin-compiler` library
- The gradle plugin uses the `ComponentRegistrar` from the `kotlin-compiler-embeddable` library

The extensions that are registered are the same so they are reused.

So for my [example project](https://github.com/Tree4Free/compiler-plugin-test-sample) I created a second copy of the `CustomRegistrar` class and moved it to the `maven-plugin` module which uses the `kotlin-compiler`

Thanks for the help :)
